src/lib.rs

```diff
@@ -0,0 +1,560 @@
+            self.left_rotation(index);
+        }
+    }
+
+    fn left_right_rotation(&mut self, index: usize) {
+        if let Some(left_index) = self.nodes[index].left_child {
+            self.left_rotation(left_index);
+            self.right_rotation(index);
+        }
+    }
+
+    fn rebalance(&mut self, index: usize) {
+        let balance_factor = self.balance_factor(&self.nodes[index]);
+        if balance_factor == -2 {
+            // right-heavy tree
+            let right_index = self.nodes[index].right_child.unwrap();
+            if self.balance_factor(&self.nodes[right_index]) == -1 {
+                self.left_rotation(index);
+            } else {
+                self.right_left_rotation(index);
+            }
+        } else if balance_factor == 2 {
+            // left-heavy
+            let left_index = self.nodes[index].left_child.unwrap();
+            if self.balance_factor(&self.nodes[left_index]) == 1 {
+                self.right_rotation(index);
+            } else {
+                self.left_right_rotation(index);
+            }
+        } else if balance_factor < -2 || balance_factor > 2 {
+            panic!("Invalid balance factor for {}: {}", index, balance_factor);
+        }
+    }
+
+    fn is_balanced(&self, index: usize) -> bool {
+        let balance_factor = self.balance_factor(&self.nodes[index]);
+        balance_factor >= -1 && balance_factor <= 1
+    }
+
+    /// Return an iterator over the values in this tree.
+    ///
+    /// # Example
+    /// ```
+    /// use avl_tree::SearchTree;
+    /// let mut tree: SearchTree<u64> = SearchTree::new();
+    /// tree.insert(5);
+    /// tree.insert(2);
+    /// tree.insert(1);
+    /// assert!(tree.iter().zip(vec![1, 2, 5]).all(|(x, y)| *x == y));
+    /// ```
+    pub fn iter(&'a self) -> SearchTreeIter<'a, T> {
+        let mut start_node = None;
+        if let Some(mut leftmost_node) = self.root {
+            while let Some(node) = self.nodes[leftmost_node].left_child {
+                leftmost_node = node;
+            }
+            start_node = Some(leftmost_node);
+        }
+        SearchTreeIter {
+            nodes: &self.nodes,
+            current_node: start_node
+        }
+    }
+}
+
+pub struct SearchTreeIter<'a, T: PartialEq + PartialOrd + Clone> where T: 'a {
+    nodes: &'a Vec<Node<T>>,
+    current_node: Option<usize>,
+}
+
+impl<'a, T> Iterator for SearchTreeIter<'a, T>
+    where T: PartialEq + PartialOrd + Clone {
+    type Item = &'a T;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        match self.current_node {
+            None => None,
+            Some(node_index) => {
+                let mut node = &self.nodes[node_index];
+                let result = &node.value;
+                if let Some(right_child) = node.right_child {
+                    self.current_node = Some(right_child);
+                    node = &self.nodes[right_child];
+                    while let Some(left_child) = node.left_child {
```

Comment:
There seems to be a recurring pattern.

Lots of code wants to find the deepest leftmost node in a given sub-tree. Should that be a function?